src/query/graphite/graphite/timespec.go

```diff
@@ -108,9 +137,146 @@ func ParseTime(s string, now time.Time, absoluteOffset time.Duration) (time.Time
 		return time.Unix(n, 0).UTC(), nil
 	}
 
+	ref, offset := s, ""
+	if strings.Contains(s, "+") {
+		stringSlice := strings.Split(s, "+")
+		ref, offset = stringSlice[0], stringSlice[1]
+		offset = "+" + offset
+	} else if strings.Contains(s, "-") {
+		stringSlice := strings.Split(s, "-")
+		ref, offset = stringSlice[0], stringSlice[1]
+		offset = "-" + offset
+	}
+	err = nil
+	parsedReference, err := ParseTimeReference(ref, now)
+	parsedOffset, err := ParseOffset(offset)
+	if err == nil {
+		return parsedReference.Add(parsedOffset), err
+	}
+
 	return now, err
 }
 
+
+func ParseTimeReference(ref string, now time.Time) (time.Time, error) {
+	hour := now.Hour()
+	minute := now.Minute()
+	refDate := time.Time{}
+
+	if ref == "" || ref == "now" {
+		return now, nil
+	}
+
+	// check if the time ref fits an absolute time format
+	for _, format := range formats {
+		t, err := time.Parse(format, ref)
+		if err == nil {
+			return time.Date(t.Year(), t.Month(), t.Day(), hour, minute, 0, 0, now.Location()), nil
+		}
+	}
+
+	rawRef := ref
+
+	// Time of Day Reference
+	i := strings.Index(ref,":")
+	if 0 < i && i < 3 {
+		newHour, err := strconv.ParseInt(ref[:i], 10, 0)
+		if err != nil {
+			return time.Time{}, err
+		}
+		hour = int(newHour)
+		newMinute, err := strconv.ParseInt(ref[i+1:i+3], 10, 32)
```

Comment:
Should we be bounds check i+FOO?

Reply:
Good call. I added a bounds check and a new test to make sure the bounds check is working.